File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0

        lHeight = self.calHeight(root.left)       
        rHeight = self.calHeight(root.right)
        
        lDiameter = self.diameterOfBinaryTree(root.left)
        rDiameter = self.diameterOfBinaryTree(root.right)
        
        return max(lHeight + rHeight, max(lDiameter, rDiameter))
        

    def calHeight(self, root) -> int:
        if root is None:
            return 0
        
        return 1 + max(self.calHeight(root.left), self.calHeight(root.right))
        
        

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

// the deepest node in the left and the deepest node in the right 
// we can have an ans check in every node
// in every node, we take the deepest node on left and deepest on right
// calculate their distance

class Solution {
public:

    int ans = 0;
    int calculateDist(TreeNode* root) {

        if (!root)
            return 0;
        
        int leftDist = calculateDist(root->left);
        int rightDist = calculateDist(root->right);

        // cout << root->val << " " << leftDist << " " << rightDist << endl;

        ans = max(ans, leftDist + rightDist + 1);
        return 1 + max(leftDist, rightDist);
    }

    int diameterOfBinaryTree(TreeNode* root) {
        calculateDist(root);
        // by the number of edges between them
        // ans is the number of nodes
        return ans - 1;
    }
};